lib/ra10ke/git_repo.rb

```diff
@@ -0,0 +1,97 @@
+require 'tempfile'
+
+module Ra10ke
+  class GitRepo
+    attr_reader :url
+
+    REMOTE_REFS_CMD = 'git ls-remote --symref'
+    CLONE_CMD = 'git clone --no-tags'
+    SHOW_CMD = 'git show'
+
+    def initialize(url)
+      @url = url
+    end
+
+    # @return [Array] - the raw data from the git ls-remote command as lines array
+    # return empty array if url or command failed
+    def remote_refs
+      @remote_refs ||= begin
+        data, success = run_command("#{REMOTE_REFS_CMD} #{url}")
+        success ? data.lines : []
+      end
+    end
+
+    # @return [Boolean] true if the git url is valid
+    def valid_url?
+      !remote_refs.empty?
+    end
+
+    # @return [Boolean] - return true if the commit sha is valid
+    # @param url [String] - the git string either https or ssh url
+    # @param ref [String] - the sha id
+    def valid_commit?(sha)
+      return false if sha.nil? || sha.empty?
+      return true if valid_ref?(sha)
+      # cloning is a last resort if for some reason we cannot
+      # remotely get via ls-remote
+      Dir.mktmpdir do |dir|
+        run_command("#{CLONE_CMD} #{url} #{dir}", silent: true)
+        Dir.chdir(dir) do
+          _, status = run_command("#{SHOW_CMD} #{sha}", silent: true)
+          status
+        end
+      end
+    end
+
+    # @return [Boolean] - return true if the ref is valid
+      # @param url [String] - the git string either https or ssh url
+      # @param ref [String] - the ref object, branch name, tag name, or commit sha, defaults to HEAD
+      def valid_ref?(ref = 'HEAD')
+        return false if ref.nil?
+        found = all_refs.find do |data|
+          # we don't need to bother with these types
+          next if data[:type] == :pull || data[:type] == :merge_request     
+          # is the name equal to the tag or branch?  Is the commit sha equal?     
+          data[:name].eql?(ref) || data[:sha].slice(0,8).eql?(ref.slice(0,8))
+        end
+        !found.nil?
+      end
```

Comment:
Nit; This block seems to be indented wrong.

Reply:
I ran rubocop to fix this.  Will need to add rubocop support in a different PR.